### PR TITLE
Android: settings lost on GUI state change

### PR DESCRIPTION
**Pull request by pljones**
**Short description of changes**

This change implements `applicationStateChanged` and `saveStateRequest` signals to save settings.

This mostly affects mobile platforms - it was identified on Android but may also affect iOS (untested).

CHANGELOG: Android: save settings on app state change

**Context: Fixes an issue?**

Fixes: #2395

**Does this change need documentation? What needs to be documented and how?**

This behaviour hasn't been documented, so no change needed.

**Status of this Pull Request**

Tested on Android 11 (Moto G9 Power retgb) (Qt5) - fixes the problem which existed in 3.10.0rc1
Tested on Linux (Qt6) - no noticable effect

**What is missing until this pull request can be merged?**

* [ ] Testing on earlier and later supported versions of Android
* [x] Testing on supported iOS versions (pre and post fix)
* [ ] Testing on other supported platforms

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

I did see this but it appears unrelated...
```
Build for MacOS (artifacts)
...
Installing create-dmg@1.1.0
~/Library/Cache/jamulus-homebrew-bottles ~/work/jamulus/jamulus
Building fresh create-dmg@1.1.0 package
Error: Cannot extract formula to homebrew/cask!
Error: Process completed with exit code 1.
```
and for the Windows Jack build:
```
Project ERROR: Error: jack.h was not found in the expected location (C:/Program Files (x86)). Ensure that the right JACK2 variant is installed (32 Bit vs. 64 Bit).
```

AUTOBUILD: Please build all targets

### CODE REVIEW

**Comment by ann0see** [comment]
iOS didn't experience that, I think.

File: src/settings.h
Comment (ann0see) [inline src/settings.h:47]: This seems a bit odd to me. Are you sure this isn't just fixing symptoms?
Concept wise, the settings and gui shouldn't be coupled that closely.

Comment (pljones) [reply src/settings.h:47]: I guess it could be done with something like this in `main.cpp`:
```
// ... get the app object to pApp or whatever
// ... get the settings object to pSettings or whatever
#ifndef HEADLESS
        if ( bUseGUI )
        {
            const QGuiApplication* pGApp = static_cast<const QGuiApplication*> ( pApp );

            QObject::connect ( pGApp, &QGuiApplication::saveStateRequest, this, [=] ( QSessionManager& ) { pSettings->Save(); );

            QObject::connect ( pGApp, &QGuiApplication::applicationStateChanged, this, [=] ( Qt::ApplicationState state ) {
                if ( Qt::ApplicationActive != state )
                {
                    pSettings->Save();
                }
            } );
        }
#endif
```

Comment (pljones) [reply src/settings.h]: OK, tested the new approach and it worked as expected.

**Comment by ann0see** [comment]
iOS name is still saved; mute myself seems to be lost, but that could be another bug. I believe iOS was never affected.

File: src/settings.h
Comment (ann0see) [inline src/settings.h:56]: Maybe put a comment here describing what and why this is needed.

**Review by ann0see — COMMENTED** [review]
Hmm. I'd like to have someone else review this.

File: src/settings.h
Comment (pljones) [reply src/settings.h:58]: Done.

**Comment by pljones** [comment]
Windows JACK build failing
```
...
Downloading 64 Bit and 32 Bit JACK installer (if needed)...
Installing JACK2 64-bit...
64bit JACK installation completed successfully
Installing JACK2 32-bit...
32bit JACK installation completed successfully
...
Project MESSAGE: building version "3.10.0dev-6d73f1a" (intermediate in git repository)
Project MESSAGE: Using JACK.
Project MESSAGE: Native x86_64 build
Project ERROR: Error: jack.h was not found in the expected location (C:/Program Files). Ensure that the right JACK2 variant is installed (32 Bit vs. 64 Bit).
Native command C:\Qt\6.6.0\msvc2019_64\bin\qmake.exe returned with exit code 3
At D:\a\jamulus\jamulus\windows\deploy_windows.ps1:45 char:9
+         Throw "Native command $Command returned with exit code $LastE ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Native command ...ith exit code 3:String) [], RuntimeException
    + FullyQualifiedErrorId : Native command C:\Qt\6.6.0\msvc2019_64\bin\qmake.exe returned with exit code 3
 
deploy_windows.ps1 failed with exit code 1
At D:\a\jamulus\jamulus\.github\autobuild\windows.ps1:218 char:9
+         throw "deploy_windows.ps1 failed with exit code $LastExitCode ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (deploy_windows....ith exit code 1:String) [], RuntimeException
    + FullyQualifiedErrorId : deploy_windows.ps1 failed with exit code 1
 
```

**Comment by hoffie** [comment]
> Windows JACK build failing

see #3206

File: src/settings.h
Comment (softins) [inline src/settings.h:62]: I looked up the signal `QGuiApplication::saveStateRequest`, and noticed it said

**Note: You should use [Qt::DirectConnection](https://doc.qt.io/qt-5/qt.html#ConnectionType-enum) when connecting to this signal.**

I'm not sure exactly what that means, but I wanted to check you'd considered it.

**Review by softins — COMMENTED** [review]
Code looks good. Just one question given below. Will be happy to approve after that.

File: src/settings.h
Comment (softins) [reply src/settings.h:62]: OK, https://doc.qt.io/qt-5/qt.html#ConnectionType-enum shows what it means. I think since the receiver is a lambda and not a thread elsewhere, it will default to `Qt:DirectConnection` anyway. Would it be a good idea to specify it explicitly?

Comment (pljones) [reply src/settings.h]: No, I hadn't spotted that.  Also, I noticed the signal is defined inside a `#ifndef` in its header, so I wrapped the usage in the same.

**Review by softins — APPROVED** [review]
LGTM

**Comment by ann0see** [comment]
As I said, I can't test on a real device. So I can't reliably test this. 
If it's ok you can merge it without my approval.

**Comment by softins** [comment]
Just built this on RPi and tried it out. Works fine. I noticed that it now saves `Jamulus.ini` on state changes, such as moving the main window, or exiting from the settings dialog. This verifies that the new code is getting invoked. In comparison, 3.10.0 only updates `Jamulus.ini` on exiting the program.

**Comment by softins** [comment]
This change seems to break registration with a directory in GUI servers. See https://github.com/jamulussoftware/jamulus/issues/3287#issuecomment-2197726812

**Pull request merged**
